plugins/net_plugin/net_plugin.cpp

```diff
@@ -2306,17 +2294,40 @@ namespace eosio {
             send_handshakes();
          }
       } else if( state == lib_catchup ) {
-         if( blk_num >= sync_known_lib_num ) {
+         std::unique_lock g_sync( sync_mtx );
+         if( blk_applied && blk_num >= sync_known_lib_num ) {
             peer_dlog( c, "All caught up with last known last irreversible block resending handshake" );
             set_state( in_sync );
             g_sync.unlock();
             send_handshakes();
-         } else if( blk_num >= sync_last_requested_num ) {
-            request_next_chunk( std::move( g_sync) );
          } else {
-            g_sync.unlock();
-            peer_dlog( c, "calling sync_wait" );
-            c->sync_wait();
+            if (!blk_applied) {
+               if (blk_num >= c->sync_last_requested_block) {
+                  peer_dlog(c, "calling cancel_wait, block ${b}", ("b", blk_num));
+                  c->cancel_wait();
+               } else {
+                  peer_dlog(c, "calling sync_wait, block ${b}", ("b", blk_num));
+                  c->sync_wait();
+               }
+
+               sync_next_expected_num = blk_num + 1;
+            } else {
+               if (sync_last_requested_num != 0 && blk_num >= sync_last_requested_num) {
+                  sync_next_expected_num = blk_num + 1;
+                  request_next_chunk(std::move(g_sync));
+                  return;
+               }
+            }
+
+            uint32_t head = my_impl->get_chain_head_num();
+            if (head + sync_req_span > sync_last_requested_num) { // don't allow to get too far head (one sync_req_span)
+               if (sync_next_expected_num > sync_last_requested_num && sync_last_requested_num < sync_known_lib_num) {
+                  fc_dlog(logger, "Requesting range ahead, head: ${h} blk_num: ${bn} sync_next_expected_num ${nen} sync_last_requested_num: ${lrn}",
+                          ("h", head)("bn", blk_num)("nen", sync_next_expected_num)("lrn", sync_last_requested_num));
+                  request_next_chunk(std::move(g_sync));
+               }
+            }
```

Comment:
Probably this block is taking care of my comment above, but then do we need the `else` case above?